### ISSUE TITLE
Singularity at pi in so3_log_map()

### ISSUE THREAD

**Issue by juldaani** [facebookresearch/pytorch3d#188]
In addition to singularity at R=I there exist also singularity at phi = pi:

~~~~
from pytorch3d.transforms.so3 import so3_log_map
import torch

R = torch.tensor([[-0.99970424, 0.000973952, 0.024300903],
                  [ 0.000737710, -0.99752367, 0.070327967],
                  [ 0.024309222,  0.070325091, 0.99722791]]).unsqueeze(0)

ax_angle = so3_log_map(R)

print(ax_angle)
~~~~

The computation near phi = pi blows up the output.
=> ax_angle = [ 51.6741,  149.4666, 4244.7529]

**Comment by davnov134** [comment]
Unfortunately, I do not believe that the instability you mention can be avoided. The computation of the log of a near-identity matrix involves division by very tiny numbers. You can improve the robustness by tweaking the `eps` parameter of `so3_log_map`.

**Comment by juldaani** [comment]
Tweaking eps doesn't help since phi is only checked that it is not near zero. In this case phi=pi*N, where N is integer. So we have singularities when rotation angle is pi, 2*pi, 3*pi, ....  If no one's going to fix this at least it should be very clearly documented that there is singularity around phi=pi. I spent quite a while finding bug in my code because of this. Basically this makes the function useless in many use cases.

**Comment by davnov134** [comment]
Ok, you are right. I have implemented a fix that allows for log_rots containing multiples of PI. We should push the fix in the course of today.

The fix looks as follows:
```
    phi_sin = phi.sin()
    phi_denom = (
        torch.clamp(phi_sin.abs(), eps) * phi_sin.sign()
        + (phi_sin == 0).type_as(phi) * eps
    )
    log_rot_hat = (phi / (2.0 * phi_denom))[:, None, None] * (R - R.permute(0, 2, 1))
```

It passes all numeric tests that convert log_rots with multiples of Pi in them.

**Comment by juldaani** [comment]
Thank you very much. I really appreciate this and I think other users would also

**Comment by davnov134** [comment]
Hi, we pushed a fix: 
https://github.com/facebookresearch/pytorch3d/commit/34a0df0630c964d4e4be225b1dc0ccf166743e75

Please let us know whether this resolved your issue.

**Comment by juldaani** [comment]
Hey

I tried the fix but unfortunately it still seems to do strange things around pi:

~~~~
R1 = torch.tensor([[-0.99007475, -0.02610438,  0.13809596],
                              [ 0.0309009,  -0.9989872,   0.0327038 ],
                              [ 0.13710241,  0.0366465,   0.9898788 ]]).unsqueeze(0)
R2 = torch.tensor([[-0.9902569,   0.00336199,  0.13921246],
                              [ 0.00223879, -0.999195, 0.04005579],
                              [ 0.13923505, 0.03997718, 0.98945206]]).unsqueeze(0)

so3_log_map(R1, eps=0.0001)  =>  tensor([[0.2147, 0.0541, 3.1049]])
so3_log_map(R2, eps=0.0001)  =>  tensor([[-0.2529, -0.0727, -3.6134]])

matrix_to_euler_angles(R1, convention='XYZ')  =>  tensor([[-0.0330,  0.1385,  3.1152]])
matrix_to_euler_angles(R2, convention='XYZ')  =>  tensor([[-0.0405,  0.1397, -3.1382]])
~~~~

When comparing outputs between so3_log_map() and matrix_to_euler_angles() there is quite large  jump in the outputs of so3_log_map(). Running the code I'm currently developing with so3_log_map() also confirms this behaviour. 

I tried with different epsilons but it didn't help.

**Comment by davnov134** [comment]
Hi, thanks a lot for the feedback. This really helps us to make the library more useful.

Indeed, the instability around multiples of PI is hard to avoid. E.g. the lower part of the "Log map from SO(3) to so(3)" section in https://en.wikipedia.org/wiki/Axis%E2%80%93angle_representation describes the issues you are currently running into.

There might a solution to do the logmap using quaternions, which do not have this kind of instability. I will investigate more and keep you posted.

**Comment by NatureGeorge** [comment]
For people who reached out to these threads looking for solutions to the instabilities of `so3_log_map`, `pytorch3d.transforms.rotation_conversions.matrix_to_axis_angle` is a choice. It is more stable since it is a wrapper of `quaternion_to_axis_angle(matrix_to_quaternion(matrix))` utilizing the advantages of quaternions.

But it should be noted that the angle yielded by `matrix_to_axis_angle` in ranging from 0 to 2pi, thus it would require a conversion like this:

![image](https://user-images.githubusercontent.com/43134199/179930412-7d6e3447-2d56-4ffb-aa3c-675459ef0a2c.png)

```py
R = random_rotations(500)
axis_angle = matrix_to_axis_angle(R)
angle_in_2pi = axis_angle.norm(dim=1)
mask = angle_in_2pi > torch.pi
axis_angle[mask] = torch.einsum('km,k->km', axis_angle[mask], 1-2*torch.pi/angle_in_2pi[mask])
```

But I am not sure whether this conversion would affect differentiations.
